src/bvals/boundary_conditions_generic.hpp

```diff
@@ -35,7 +35,103 @@ namespace parthenon {
 namespace BoundaryFunction {
 
 enum class BCSide { Inner, Outer };
-enum class BCType { Outflow, Reflect, ConstantDeriv, Fixed, FixedFace };
+enum class BCType { Outflow, Reflect, ConstantDeriv, Fixed, FixedFace, Periodic };
+
+// TODO(BRR) add support for specific swarms?
+template <CoordinateDirection DIR, BCSide SIDE, BCType TYPE>
+void GenericSwarmBC(std::shared_ptr<Swarm> &swarm) {
+  // make sure DIR is X[123]DIR so we don't have to check again
+  static_assert(DIR == X1DIR || DIR == X2DIR || DIR == X3DIR, "DIR must be X[123]DIR");
+
+  // convenient shorthands
+  constexpr bool X1 = (DIR == X1DIR);
+  constexpr bool X2 = (DIR == X2DIR);
+  constexpr bool X3 = (DIR == X3DIR);
+  constexpr bool INNER = (SIDE == BCSide::Inner);
+
+  auto swarm_d = swarm->GetDeviceContext();
+  int max_active_index = swarm->GetMaxActiveIndex();
+
+  auto pmb = swarm->GetBlockPointer();
+
+  auto &x = swarm->Get<Real>("x").Get();
+  auto &y = swarm->Get<Real>("y").Get();
+  auto &z = swarm->Get<Real>("z").Get();
+
+  // TODO(BRR) do something about all these if statements
+  pmb->par_for(
+      PARTHENON_AUTO_LABEL, 0, max_active_index, KOKKOS_LAMBDA(const int n) {
+        if (swarm_d.IsActive(n)) {
+          if (X1) {
+            if (INNER) {
+              if (TYPE == BCType::Periodic) {
```

Comment:
I think because it's a template parameter you can make these all `if constexpr` and then the if statements are just fine.

Reply:
Made that change (in the fields version of the generic BC as well) thanks for the suggestion I think `if constexpr` should work nicely here, the alternative would be verbose.

Reply:
I was not able to switch the field version of `GenericBC` to use `constexpr if` -- the nvcc compiler errors ultimately weren't making sense so I'm worried the compiler just isn't stable enough to support `constexpr if` on a function that intricate. Maybe that function can be revisited in a subsequent PR.

Reply:
seems like this is resolved.